### PR TITLE
Judges | Assign cases | Assign to 'Other'

### PR DESCRIPTION
**Pull request by evankroske**
- Fixes https://github.com/department-of-veterans-affairs/caseflow/issues/5174

### Description
Allows judges to assign cases to any attorney at the board.

### Testing Plan
- [ ] Become BVARERDMAN.
- [ ] Assign a case to Lela Shields.
- [ ] Become BVAAABSHIRE.
- [ ] Confirm that the case is assigned to Lela.
- [ ] On "Cases to Assign", assign an unassigned case to Jazlyn Schimmel.
- [ ] Become BVARERDMAN.
- [ ] Confirm that the case has been assigned to Jazlyn.

### CODE REVIEW

File: client/app/queue/QueueActions.js [hunk 1]
```diff
@@ -249,3 +249,27 @@ export const reassignTasksToUser = ({ tasks, assigneeId, previousAssigneeId }) =
...
+    error
+  }
+});
+
+export const fetchAllAttorneys = () => (dispatch) => {
+  return ApiUtil.get(
+    '/users?role=Attorney').
+    then((resp) => resp.body).
+    then(
+      (resp) => dispatch(receiveAllAttorneys(resp.attorneys))).
```
Comment (dannysellers) [inline client/app/queue/QueueActions.js]: Total nitpick: This one line is formatted differently than the others (as is the first one, actually)

File: client/app/queue/reducers.js [hunk 2]
```diff
@@ -38,7 +38,8 @@ export const initialState = {
   },
   attorneysOfJudge: [],
   tasksAndAppealsOfAttorney: {},
-  isTaskAssignedToUserSelected: {}
+  isTaskAssignedToUserSelected: {},
+  allAttorneys: {}
```
Comment (dannysellers) [inline client/app/queue/reducers.js]: Why not just `attorneys`? We don't have `allJudges`

File: client/app/queue/reducers.js [hunk 3]
```diff
@@ -359,6 +360,22 @@ const workQueueReducer = (state = initialState, action = {}) => {
...
   }
+  case ACTIONS.RECEIVE_ALL_ATTORNEYS:
+    return update(state, {
+      allAttorneys: {
+        $set: {
+          data: action.payload.attorneys
+        }
+      }
+    });
+  case ACTIONS.ERROR_ALL_ATTORNEYS:
```
Comment (dannysellers) [inline client/app/queue/reducers.js]: I feel like reducer actions should be semantically tied to the action they're related to. `ERROR_ALL_ATTORNEYS` doesn't have any kind of verb / indication that it's related to loading, rather than something like form validation. Why not name it something like `ERROR_LOADING_ATTORNEYS`/`ERROR_LOADING_ALL_ATTORNEYS`?

File: client/app/queue/JudgeAssignTaskListView.jsx [hunk 4]
```diff
@@ -58,17 +58,14 @@ class JudgeAssignTaskListView extends React.PureComponent {
...
-    };
-
-    return ApiUtil.get(`/users?role=Attorney&judge_css_id=${this.props.userCssId}`, requestOptions).
+    return this.props.fetchAllAttorneys().
       then(
-        (response) => {
-          const resp = JSON.parse(response.text);
+        (attorneys) => {
+          const attorneysOfJudgeWithCssId = _.groupBy(attorneys, (attorney) => attorney.judge_css_id);
+          const attorneysOfJudge = attorneysOfJudgeWithCssId[this.props.userCssId];
```
Comment (dannysellers) [inline client/app/queue/JudgeAssignTaskListView.jsx]: Nit: why not just filter? `attorneys.filter(attorney) => attorney.judge_css_id === this.props.userCssId)`

File: client/app/queue/components/AssignWidget.jsx [hunk 5]
```diff
@@ -84,9 +91,50 @@ class AssignWidget extends React.PureComponent<Props> {
...
+        }
+        options.push({
+          label: `${judgeName}:`,
+          value: judgeCssId,
+          disabled: true
+        });
+        options.push(...attorneysOfJudgeWithCssId[judgeCssId].map(optionFromAttorney));
+      }
+      options.push({ label: ASSIGN_WIDGET_OTHER,
+        value: OTHER });
```
Comment (dannysellers) [inline client/app/queue/components/AssignWidget.jsx]: Why `OTHER` and not just `ASSIGN_WIDGET_OTHER.toUpperCase()`?

File: client/app/queue/components/AssignWidget.jsx [hunk 6]
```diff
@@ -104,7 +152,7 @@ class AssignWidget extends React.PureComponent<Props> {
           searchable
           options={options}
           placeholder="Select a user"
-          onChange={(option) => this.props.setSelectedAssignee({ assigneeId: option.value })}
+          onChange={handleChange}
```
Comment (dannysellers) [inline client/app/queue/components/AssignWidget.jsx]: Nit: why extract this callback / move it to a different part of the file? It looks like this is the only place it's used, but now it's separated by 50 lines of indirectly-related / unrelated code

File: client/app/queue/components/AssignWidget.jsx [hunk 7]
```diff
@@ -84,9 +91,50 @@ class AssignWidget extends React.PureComponent<Props> {
...
+      return <Alert type="error" title="Error fetching attorneys" message="Please refresh the page." />;
+    }
+    const optionFromAttorney = (attorney) => ({ label: attorney.full_name,
+      value: attorney.id.toString() });
+    const handleChange = (option) => this.props.setSelectedAssignee({ assigneeId: option.value });
+    const attorneys = allAttorneys.data;
+
+    const options = [];
+
+    {
```
Comment (dannysellers) [inline client/app/queue/components/AssignWidget.jsx]: Why use a block here? It doesn't look like there're any variables declared inside it that need a separate scope, and if it's just a visual/semantic distinction, it should probably just go in its own function

File: client/app/queue/components/AssignWidget.jsx [hunk 8]
```diff
@@ -84,9 +91,50 @@ class AssignWidget extends React.PureComponent<Props> {
...
+      const attorneysOfJudgeWithCssId = _.groupBy(attorneys, (attorney) => attorney.judge_css_id);
+      const judgeCssIds =
+        [userCssId].concat(Object.keys(attorneysOfJudgeWithCssId).filter((cssId) => cssId !== userCssId));
```
Comment (dannysellers) [inline client/app/queue/components/AssignWidget.jsx]: I don't think we need to handle `userCssId` separately/specially here—wouldn't it already be included in the set of all judges? Why not something like `judgeCssIds = _.uniq(_.map(attorneys, 'judge_css_id'))`?

File: client/app/queue/components/AssignWidget.jsx [hunk 9]
```diff
@@ -83,10 +90,60 @@ class AssignWidget extends React.PureComponent<Props> {
...
+    }
+
+    const options = [];
+
+    options.push({
+      label: `${this.fullNameOfJudgeWithCssId(judgeCssId) || 'Other'}:`,
+      value: judgeCssId,
+      disabled: true
+    });
+    options.push(...attorneysOfJudgeWithCssId[judgeCssId].map((attorney) => ({ label: attorney.full_name,
```
Comment (dannysellers) [inline client/app/queue/components/AssignWidget.jsx]: Thank you for moving these into separate functions. It looks like `options` might not be needed any more here—could it just return the result of this map?

File: client/app/queue/components/AssignWidget.jsx [hunk 10]
```diff
@@ -83,10 +90,60 @@ class AssignWidget extends React.PureComponent<Props> {
...
+
+    {
+      const attorneysOfJudgeWithCssId = _.groupBy(attorneys, (attorney) => attorney.judge_css_id);
+      const judgeCssIds =
+        [userCssId].concat(Object.keys(attorneysOfJudgeWithCssId).filter((cssId) => cssId !== userCssId));
+
+      for (const judgeCssId of judgeCssIds) {
+        options.push(...this.optionsFromJudgeAndAttorneys(judgeCssId, attorneysOfJudgeWithCssId));
+      }
+      options.push({ label: ASSIGN_WIDGET_OTHER,
```
Comment (dannysellers) [inline client/app/queue/components/AssignWidget.jsx]: It looks like the textbox described in the AC is missing—this just would assign to Other
> Verify that when the user chooses 'Other' from the drop down, they have a text box to enter the name of an attorney or judge.

File: client/app/queue/components/AssignWidget.jsx [hunk 11]
```diff
@@ -83,10 +90,60 @@ class AssignWidget extends React.PureComponent<Props> {
...
+        value: OTHER });
```
Comment (dannysellers) [inline client/app/queue/components/AssignWidget.jsx]: Why use a different constant? `ASSIGN_WIDGET_OTHER.toUpperCase()` should work

File: client/app/queue/components/AssignWidget.jsx [hunk 12]
```diff
@@ -83,10 +90,60 @@ class AssignWidget extends React.PureComponent<Props> {
...
+
+    const options = [];
+
+    {
+      const attorneysOfJudgeWithCssId = _.groupBy(attorneys, (attorney) => attorney.judge_css_id);
+      const judgeCssIds =
+        [userCssId].concat(Object.keys(attorneysOfJudgeWithCssId).filter((cssId) => cssId !== userCssId));
+
+      for (const judgeCssId of judgeCssIds) {
+        options.push(...this.optionsFromJudgeAndAttorneys(judgeCssId, attorneysOfJudgeWithCssId));
```
Comment (dannysellers) [inline client/app/queue/components/AssignWidget.jsx]: Nit: It looks like we could build `options` with a `map` at this point

File: client/app/queue/components/AssignWidget.jsx [hunk 13]
```diff
@@ -83,10 +90,60 @@ class AssignWidget extends React.PureComponent<Props> {
...
   render = () => {
-    const { attorneysOfJudge, selectedAssignee, error, success } = this.props;
-    const options = attorneysOfJudge.map((attorney) => ({ label: attorney.full_name,
-      value: attorney.id.toString() }));
+    const { selectedAssignee, error, success, allAttorneys, userCssId } = this.props;
+
+    if (!allAttorneys.data && !allAttorneys.error) {
+      return <SmallLoader message="Loading..." spinnerColor={LOGO_COLORS.QUEUE.ACCENT} />;
+    }
+    if (allAttorneys.error) {
```
Comment (dannysellers) [inline client/app/queue/components/AssignWidget.jsx]: Nit: If this was moved above the check for `!data && !error`, we could remove the `!error` check
```javascript
if (error) {
  return
}
if (!data) {
  return
}
continue
```

File: client/app/queue/components/AssignWidget.jsx (hunk 8 above)
Comment (evankroske) [reply client/app/queue/components/AssignWidget.jsx]: The user's attorneys needs to come first so that the user can easily find their attorneys when there are a lot of attorneys in the dropdown. This code just puts the current user's CSS ID first in the list so that there attorneys are first.

File: client/app/queue/AssignedCasesPage.jsx [hunk 14]
```diff
@@ -63,7 +63,8 @@ class AssignedCasesPage extends React.PureComponent {
       {featureToggles.judge_assign_cases &&
         <AssignWidget
           previousAssigneeId={attorneyId}
-          onTaskAssignment={(params) => props.reassignTasksToUser(params)} />}
+          onTaskAssignment={(params) => props.reassignTasksToUser(params)}
+          userCssId={this.props.userCssId} />}
```
[the remark this answers is not in the document]
Comment (evankroske) [reply client/app/queue/AssignedCasesPage.jsx]: Makes sense. I'll make it global

File: client/app/queue/JudgeAssignTaskListView.jsx [hunk 15]
```diff
@@ -127,12 +124,14 @@ class JudgeAssignTaskListView extends React.PureComponent {
...
-                userId={this.props.userId.toString()} />}
+                userId={this.props.userId.toString()}
+                userCssId={this.props.userCssId} />}
           />
           <PageRoute
             path={`${match.url}/:attorneyId`}
             title="Assigned Cases | Caseflow"
-            component={AssignedCasesPage}
+            render={
+              ({ match: innerMatch }) => <AssignedCasesPage userCssId={this.props.userCssId} match={innerMatch} />}
```
[the remark this answers is not in the document]
Comment (evankroske) [reply client/app/queue/JudgeAssignTaskListView.jsx]: Sounds good! I'll make this change

File: client/app/queue/QueueActions.js (hunk 1 above)
Comment (evankroske) [reply client/app/queue/QueueActions.js]: OK, I'll fix it

File: client/app/queue/reducers.js (hunk 2 above)
Comment (evankroske) [reply client/app/queue/reducers.js]: Makes sense! I'll rename this field

File: client/app/queue/reducers.js (hunk 3 above)
Comment (evankroske) [reply client/app/queue/reducers.js]: OK, I'll rename it

File: client/app/queue/JudgeAssignTaskListView.jsx (hunk 4 above)
Comment (evankroske) [reply client/app/queue/JudgeAssignTaskListView.jsx]: Good point! I'll change it

File: client/app/queue/components/AssignWidget.jsx (hunk 13 above)
Comment (evankroske) [reply client/app/queue/components/AssignWidget.jsx]: Good point! I'll change it

File: client/app/queue/components/AssignWidget.jsx (hunk 7 above)
Comment (evankroske) [reply client/app/queue/components/AssignWidget.jsx]: This was left over from a previous version of the code. I'll get rid of it

File: client/app/queue/components/AssignWidget.jsx (hunk 10 above)
Comment (evankroske) [reply client/app/queue/components/AssignWidget.jsx]: Thanks for pointing this out! I forgot to remove this option. I talked to the designers, and we decided to just put all of the attorneys in one dropdown instead of adding a second dropdown. This code was left over from an earlier version.

File: client/app/queue/components/AssignWidget.jsx (hunk 5 above)
Comment (evankroske) [reply client/app/queue/components/AssignWidget.jsx]: OK, I'll change this

File: client/app/queue/components/AssignWidget.jsx (hunk 11 above)
Comment (evankroske) [reply client/app/queue/components/AssignWidget.jsx]: OK, I'll change this

File: client/app/queue/components/AssignWidget.jsx (hunk 6 above)
Comment (evankroske) [reply client/app/queue/components/AssignWidget.jsx]: This was left over from an earlier version. I'll change it back
